cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java

```diff
@@ -241,49 +248,105 @@ public void setConfig(HttpRequest request, HttpResponder responder,
-        return;
-      }
+      currConfig = streamAdmin.getConfig(stream);
+    } catch (IOException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, "Stream " + stream + " does not exist.");
+      return;
+    }
 
-      streamAdmin.updateConfig(config);
-      responder.sendStatus(HttpResponseStatus.OK);
+    StreamConfig requestedConfig = getAndValidateConfig(currConfig, request, responder);
+    // null is returned if the requested config is invalid. An appropriate response will have already been written
+    // to the responder so we just need to return.
+    if (requestedConfig == null) {
+      return;
+    }
 
-    } catch (IOException e) {
-      responder.sendString(HttpResponseStatus.NOT_FOUND, "Stream does not exists");
+    streamAdmin.updateConfig(requestedConfig);
+    // if the schema has changed, we need to recreate the hive table. Changes in format and settings don't require
+    // a hive change, as they are just properties used by the stream storage handler.
+    Schema currSchema = currConfig.getFormat().getSchema();
+    Schema newSchema = requestedConfig.getFormat().getSchema();
+    if (exploreEnabled && !currSchema.equals(newSchema)) {
+      exploreFacade.disableExploreStream(stream);
+      exploreFacade.enableExploreStream(stream);
     }
-  }
 
-  private String getMetricsContext() {
-    return Constants.Gateway.METRICS_CONTEXT + "." + cConf.getInt(Constants.Stream.CONTAINER_INSTANCE_ID, 0);
+    responder.sendStatus(HttpResponseStatus.OK);
   }
 
+  // given the current config for a stream and requested config for a stream, get the new stream config
+  // with defaults in place of missing settings, and validation performed on the requested fields.
+  // If a field is missing or invalid, this method will write an appropriate response to the responder and
+  // return a null config.
+  private StreamConfig getAndValidateConfig(StreamConfig currConfig, HttpRequest request, HttpResponder responder) {
+    // get new config settings from the request. Only TTL and format spec can be changed, which is
+    // why the StreamProperties object is used instead of a StreamConfig object.
+    Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()));
+    StreamProperties properties;
+    try {
+      properties = GSON.fromJson(reader, StreamProperties.class);
+    } catch (Exception e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Invalid stream configuration. Please check that the " +
+        "configuration is a valid JSON Object with a valid schema.");
+      return null;
+    }
 
-  private StreamConfig getConfigUpdate(HttpRequest request, StreamConfig config) {
-    JsonObject json = GSON.fromJson(request.getContent().toString(Charsets.UTF_8), JsonObject.class);
+    // if no ttl is given, use the existing ttl.
+    Long newTTL = properties.getTTL();
+    if (newTTL == null) {
```

Comment:
I still need the if else because it needs to be converted to millis if it comes from the request, but not if it comes from the existing config.